test/jdk/java/net/URLConnection/RequestProperties.java

```diff
@@ -23,57 +23,186 @@
+ * @run testng RequestProperties
  */
 
-import java.net.*;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
 
 public class RequestProperties {
-    static int failed;
 
-    public static void main (String args[]) throws Exception {
-        List<String> urls = new ArrayList<>();
+    @DataProvider(name = "urls")
+    private Object[][] urls() {
+        final List<String> urls = new ArrayList<>();
         urls.add("http://foo.com/bar/");
         urls.add("jar:http://foo.com/bar.html!/foo/bar");
         urls.add("file:/etc/passwd");
-        if (hasFtp())
+        if (hasFtp()) {
             urls.add("ftp://foo:bar@foobar.com/etc/passwd");
+        }
+        final Object[][] data = new Object[urls.size()][1];
+        for (int i = 0; i < urls.size(); i++) {
+            data[i][0] = urls.get(i);
+        }
+        return data;
+    }
 
-        for (String urlStr : urls)
-            test(new URL(urlStr));
 
-        if (failed != 0)
-            throw new RuntimeException(failed + " errors") ;
+    /**
+     * Test that {@link java.net.URLConnection#setRequestProperty(String, String)} throws
+     * a {@link NullPointerException} when passed null key
+     */
+    @Test(dataProvider = "urls")
+    public void testSetRequestPropertyNullPointerException(final String url) throws Exception {
+        final URLConnection conn = new URL(url).openConnection();
+        try {
+            conn.setRequestProperty(null, "bar");
+            Assert.fail("setRequestProperty on " + conn.getClass().getName()
+                    + " for " + conn.getURL() + " was expected to throw"
+                    + " NullPointerException, but didn't");
+        } catch (NullPointerException npe) {
+            // expected
+        }
+        // expected to pass
+        conn.setRequestProperty("key", null);
     }
 
-    static void test(URL url) throws Exception {
-        URLConnection urlc = url.openConnection();
+    /**
+     * Test that {@link java.net.URLConnection#addRequestProperty(String, String)} throws
+     * a {@link NullPointerException} when passed null key
+     */
+    @Test(dataProvider = "urls")
+    public void testAddRequestPropertyNullPointerException(final String url) throws Exception {
+        final URLConnection conn = new URL(url).openConnection();
         try {
-            urlc.setRequestProperty(null, null);
-            System.out.println(url.getProtocol()
-                               + ": setRequestProperty(null,) did not throw NPE");
-            failed++;
-        } catch (NullPointerException e) { /* Expected */ }
+            conn.addRequestProperty(null, "hello");
+            Assert.fail("addRequestProperty on " + conn.getClass().getName()
```

Comment:
Now that the test is using TestNG, you can use the convenient Assert.assertThrows / Assert.expectThrows, e.g. something like:

static final Class<NullPointerException> NPE = NullPointerException.class;

assertThrows(NPE, () -> conn.addRequestProperty(null, "hello"));

similar for IllegalStateException

Reply:
Thank you. I wasn't aware of those APIs in TestNG. Looks like they got introduced in 6.9.5 of TestNG.
Not sure what version of TestNG, the `jtreg` tool supports. But when I changed this test to use those APIs and run it, it worked fine and passed. So I guess it's supported.
Using `jtreg -version` doesn't show the TestNG version (unlike what the docs claim), but that anyway isn't something relevant here and is a question to discuss on the jtreg mailing list:

```
jtreg -version
jtreg, version 5.1 dev 827
Installed in /home/me/jtreg/jtreg/lib/jtreg.jar
Running on platform version 1.8.0_265 from /home/me/java/Contents/Home/jre.
Built with Java(TM) 2 SDK, Version 1.8.0_172-b11 on June 21, 2020.
Copyright (c) 1999, 2019, Oracle and/or its affiliates. All rights reserved.
Use is subject to license terms.
JT Harness, version 6.0 ea b11 (June 21, 2020)
JCov 3.0-2
TestNG (testng.jar): version unknown
TestNG (jcommander.jar): version 1.72
Java Assembler Tools, version 7.0 beta b02 (June 21, 2020)
```
I've updated this PR to include your suggested change.